Make http request more robust

michellewen1516 opened the pull request:
I got the error of "failed because connect timed out (skinny.http.HTTP$)" during test. In this case, applied an empty acls rules and remove all the existing rules on Kafka. Found that enableThrowingIOException is false by default, which may not throw exception in some case. 
I wasn't able to reproduce this error, but I think the change in this PR can make it more robust.

simplesteph commented:
This is really weird, because in case of HTTP exception, we don't apply ACL:
https://github.com/simplesteph/kafka-security-manager/blob/c0674026b76e00b664fb4470c42ac3e9e8933086/src/main/scala/com/github/simplesteph/ksm/AclSynchronizer.scala#L108 

And in case of non 200 result (non `Some()`), an error is attached to the response and ACLs are not applied too:
https://github.com/simplesteph/kafka-security-manager/blob/c0674026b76e00b664fb4470c42ac3e9e8933086/src/main/scala/com/github/simplesteph/ksm/AclSynchronizer.scala#L98


If you can find some logs, and reproduce the error, I would be grateful...

michellewen1516 commented:
1. This is the log of this weird case:
Feb 20 07:45:09 ip-172-31-11-158.ec2.internal kafka-security-manager[2341]: [2020-02-20 07:45:09,692] INFO No ACL changes (AclSynchronizer)
Feb 20 07:45:20 ip-172-31-11-158.ec2.internal kafka-security-manager[2341]: [2020-02-20 07:45:20,602] WARN Method(GET) https://api.bitbucket.org/2.0/repositories/xxx_repo/acls.csv failed because connect timed out (skinny.http.HTTP$)
Feb 20 07:45:39 ip-172-31-11-158.ec2.internal kafka-security-manager[2341]: [2020-02-20 07:45:39,679] INFO No ACL changes (AclSynchronizer)
Feb 20 07:45:49 ip-172-31-11-158.ec2.internal kafka-security-manager[2341]: [2020-02-20 07:45:49,686] INFO No ACL changes (AclSynchronizer)
Feb 20 07:45:59 ip-172-31-11-158.ec2.internal kafka-security-manager[2341]: [2020-02-20 07:45:59,698] INFO No ACL changes (AclSynchronizer)

2. I did get some normal error case and the log is:
Feb 20 18:31:54 ip-172-31-11-158.ec2.internal kafka-security-manager[2341]: [2020-02-20 18:31:54,612] WARN Method(GET) https://api.bitbucket.org/2.0/repositories/xxx_repo/acls.csv failed because Read timed out (skinny.http.HTTP$)
Feb 20 18:31:54 ip-172-31-11-158.ec2.internal kafka-security-manager[2341]: [2020-02-20 18:31:54,612] ERROR Refreshing the source failed (AclSynchronizer)
Feb 20 18:31:54 ip-172-31-11-158.ec2.internal kafka-security-manager[2341]: java.net.SocketTimeoutException: Read timed out
Feb 20 18:31:54 ip-172-31-11-158.ec2.internal kafka-security-manager[2341]: at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
Feb 20 18:31:54 ip-172-31-11-158.ec2.internal kafka-security-manager[2341]: at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
Feb 20 18:31:54 ip-172-31-11-158.ec2.internal kafka-security-manager[2341]: at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
Feb 20 18:31:54 ip-172-31-11-158.ec2.internal kafka-security-manager[2341]: at java.lang.reflect.Constructor.newInstance(Constructor.java:423)
Feb 20 18:31:54 ip-172-31-11-158.ec2.internal kafka-security-manager[2341]: at sun.net.www.protocol.http.HttpURLConnection$10.run(HttpURLConnection.java:1950)
Feb 20 18:31:54 ip-172-31-11-158.ec2.internal kafka-security-manager[2341]: at sun.net.www.protocol.http.HttpURLConnection$10.run(HttpURLConnection.java:1945)
Feb 20 18:31:54 ip-172-31-11-158.ec2.internal kafka-security-manager[2341]: at java.security.AccessController.doPrivileged(Native Method)
Feb 20 18:31:54 ip-172-31-11-158.ec2.internal kafka-security-manager[2341]: at sun.net.www.protocol.http.HttpURLConnection.getChainedException(HttpURLConnection.java:1944)
Feb 20 18:31:54 ip-172-31-11-158.ec2.internal kafka-security-manager[2341]: at sun.net.www.protocol.http.HttpURLConnection.getInputStream0(HttpURLConnection.java:1514)
Feb 20 18:31:54 ip-172-31-11-158.ec2.internal kafka-security-manager[2341]: at sun.net.www.protocol.http.HttpURLConnection.getInputStream(HttpURLConnection.java:1498)
Feb 20 18:31:54 ip-172-31-11-158.ec2.internal kafka-security-manager[2341]: at java.net.HttpURLConnection.getResponseCode(HttpURLConnection.java:480)
Feb 20 18:31:54 ip-172-31-11-158.ec2.internal kafka-security-manager[2341]: at sun.net.www.protocol.https.HttpsURLConnectionImpl.getResponseCode(HttpsURLConnectionImpl.java:352)
Feb 20 18:31:54 ip-172-31-11-158.ec2.internal kafka-security-manager[2341]: at skinny.http.HTTP.request(HTTP.scala:181)
Feb 20 18:31:54 ip-172-31-11-158.ec2.internal kafka-security-manager[2341]: at skinny.http.HTTP.get(HTTP.scala:49)
Feb 20 18:31:54 ip-172-31-11-158.ec2.internal kafka-security-manager[2341]: at com.github.simplesteph.ksm.source.BitbucketCloudSourceAcl.refresh(BitbucketCloudSourceAcl.scala:57)
Feb 20 18:31:54 ip-172-31-11-158.ec2.internal kafka-security-manager[2341]: at com.github.simplesteph.ksm.AclSynchronizer.$anonfun$run$1(AclSynchronizer.scala:76)
Feb 20 18:31:54 ip-172-31-11-158.ec2.internal kafka-security-manager[2341]: at scala.util.Try$.apply(Try.scala:213)
Feb 20 18:31:54 ip-172-31-11-158.ec2.internal kafka-security-manager[2341]: at com.github.simplesteph.ksm.AclSynchronizer.run(AclSynchronizer.scala:76)
Feb 20 18:31:54 ip-172-31-11-158.ec2.internal kafka-security-manager[2341]: at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
Feb 20 18:31:54 ip-172-31-11-158.ec2.internal kafka-security-manager[2341]: at java.util.concurrent.FutureTask.runAndReset(FutureTask.java:308)
Feb 20 18:31:54 ip-172-31-11-158.ec2.internal kafka-security-manager[2341]: at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$301(ScheduledThreadPoolExecutor.java:180)
Feb 20 18:31:54 ip-172-31-11-158.ec2.internal kafka-security-manager[2341]: at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:294)
Feb 20 18:31:54 ip-172-31-11-158.ec2.internal kafka-security-manager[2341]: at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
Feb 20 18:31:54 ip-172-31-11-158.ec2.internal kafka-security-manager[2341]: at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
Feb 20 18:31:54 ip-172-31-11-158.ec2.internal kafka-security-manager[2341]: at java.lang.Thread.run(Thread.java:748)
Feb 20 18:31:54 ip-172-31-11-158.ec2.internal kafka-security-manager[2341]: Caused by: java.net.SocketTimeoutException: Read timed out
Feb 20 18:31:54 ip-172-31-11-158.ec2.internal kafka-security-manager[2341]: at java.net.SocketInputStream.socketRead0(Native Method)
Feb 20 18:31:54 ip-172-31-11-158.ec2.internal kafka-security-manager[2341]: at java.net.SocketInputStream.socketRead(SocketInputStream.java:116)
Feb 20 18:31:54 ip-172-31-11-158.ec2.internal kafka-security-manager[2341]: at java.net.SocketInputStream.read(SocketInputStream.java:171)
Feb 20 18:31:54 ip-172-31-11-158.ec2.internal kafka-security-manager[2341]: at java.net.SocketInputStream.read(SocketInputStream.java:141)
Feb 20 18:31:54 ip-172-31-11-158.ec2.internal kafka-security-manager[2341]: at sun.security.ssl.InputRecord.readFully(InputRecord.java:465)
Feb 20 18:31:54 ip-172-31-11-158.ec2.internal kafka-security-manager[2341]: at sun.security.ssl.InputRecord.read(InputRecord.java:503)
Feb 20 18:31:54 ip-172-31-11-158.ec2.internal kafka-security-manager[2341]: at sun.security.ssl.SSLSocketImpl.readRecord(SSLSocketImpl.java:975)
Feb 20 18:31:54 ip-172-31-11-158.ec2.internal kafka-security-manager[2341]: at sun.security.ssl.SSLSocketImpl.readDataRecord(SSLSocketImpl.java:933)
Feb 20 18:31:54 ip-172-31-11-158.ec2.internal kafka-security-manager[2341]: at sun.security.ssl.AppInputStream.read(AppInputStream.java:105)
Feb 20 18:31:54 ip-172-31-11-158.ec2.internal kafka-security-manager[2341]: at java.io.BufferedInputStream.fill(BufferedInputStream.java:246)
Feb 20 18:31:54 ip-172-31-11-158.ec2.internal kafka-security-manager[2341]: at java.io.BufferedInputStream.read1(BufferedInputStream.java:286)
Feb 20 18:31:54 ip-172-31-11-158.ec2.internal kafka-security-manager[2341]: at java.io.BufferedInputStream.read(BufferedInputStream.java:345)
Feb 20 18:31:54 ip-172-31-11-158.ec2.internal kafka-security-manager[2341]: at sun.net.www.http.HttpClient.parseHTTPHeader(HttpClient.java:735)
Feb 20 18:31:54 ip-172-31-11-158.ec2.internal kafka-security-manager[2341]: at sun.net.www.http.HttpClient.parseHTTP(HttpClient.java:678)
Feb 20 18:31:54 ip-172-31-11-158.ec2.internal kafka-security-manager[2341]: at sun.net.www.protocol.http.HttpURLConnection.getInputStream0(HttpURLConnection.java:1593)
Feb 20 18:31:54 ip-172-31-11-158.ec2.internal kafka-security-manager[2341]: at sun.net.www.protocol.http.HttpURLConnection.getInputStream(HttpURLConnection.java:1498)
Feb 20 18:31:54 ip-172-31-11-158.ec2.internal kafka-security-manager[2341]: at sun.net.www.protocol.https.HttpsURLConnectionImpl.getInputStream(HttpsURLConnectionImpl.java:268)
Feb 20 18:31:54 ip-172-31-11-158.ec2.internal kafka-security-manager[2341]: at skinny.http.HTTP.request(HTTP.scala:166)

I tried many cases but failed to reproduce the weird thing. I am a newbie of scala and will be grateful if you can give me some info about how to reproduce this bug. Thanks

simplesteph commented:
This is not weird:
```
Feb 20 07:45:09 ip-172-31-11-158.ec2.internal kafka-security-manager[2341]: [2020-02-20 07:45:09,692] INFO No ACL changes (AclSynchronizer)
Feb 20 07:45:20 ip-172-31-11-158.ec2.internal kafka-security-manager[2341]: [2020-02-20 07:45:20,602] WARN Method(GET) https://api.bitbucket.org/2.0/repositories/xxx_repo/acls.csv failed because connect timed out (skinny.http.HTTP$)
Feb 20 07:45:39 ip-172-31-11-158.ec2.internal kafka-security-manager[2341]: [2020-02-20 07:45:39,679] INFO No ACL changes (AclSynchronizer)
Feb 20 07:45:49 ip-172-31-11-158.ec2.internal kafka-security-manager[2341]: [2020-02-20 07:45:49,686] INFO No ACL changes (AclSynchronizer)
Feb 20 07:45:59 ip-172-31-11-158.ec2.internal kafka-security-manager[2341]: [2020-02-20 07:45:59,698] INFO No ACL changes (AclSynchronizer)
```

It just says there was an exception when trying to get a repo because of a time out, but it definitely didn't wipe out your ACL and applied none. The code I wrote shouldn't do that. If there was ACL being applied, it should show up in your log, but it doesn't. So everything works as expected as far as I can tell from your logs

michellewen1516 commented:
More info, I sent the changes to the slack channel, so you couldn't see any ADDED or REMOVED info here, but between the below two lines of log, I got REMOVED all acl rules and then ADDED ALL again.


Feb 20 07:45:20 ip-172-31-11-158.ec2.internal kafka-security-manager[2341]: [2020-02-20 07:45:20,602] WARN Method(GET) https://api.bitbucket.org/2.0/repositories/xxx_repo/acls.csv failed because connect timed out (skinny.http.HTTP$)
Feb 20 07:45:39 ip-172-31-11-158.ec2.internal kafka-security-manager[2341]: [2020-02-20 07:45:39,679] INFO No ACL changes (AclSynchronizer)

simplesteph commented:
Hmmm.... do you have a screenshot of the Slack channel?
This is really off...

michellewen1516 commented:
Yes, I have. 
Something like this:
1: 
FEB 20TH, 02:45:20 AM
REMOVED User: xxxxxxxxxxxxxx

total 20 lines of above stuff

2:
Then at 02:45:29AM
I got:
ADDED User:xxxxxxxxxxxxxxxxxxxxxx
total 20 lines of above stuff

michellewen1516 commented:
It happened almost 10 times since Feb 20th. From my tests, I always get the similar log with the normal error case. I think the ksm should throw error when:
Feb 20 07:45:20 ip-172-31-11-158.ec2.internal kafka-security-manager[2341]: [2020-02-20 07:45:20,602] WARN Method(GET) https://api.bitbucket.org/2.0/repositories/xxx_repo/acls.csv failed because connect timed out (skinny.http.HTTP$)

Can you please explain a little bit why it didn't throw any error?

simplesteph commented:
You are getting a WARN:
https://github.com/simplesteph/kafka-security-manager/blob/master/src/main/scala/com/github/simplesteph/ksm/source/BitbucketCloudSourceAcl.scala#L87 

Which returns this:
https://github.com/simplesteph/kafka-security-manager/blob/master/src/main/scala/com/github/simplesteph/ksm/source/BitbucketCloudSourceAcl.scala#L88-L92

This result is then going into 
https://github.com/simplesteph/kafka-security-manager/blob/master/src/main/scala/com/github/simplesteph/ksm/AclSynchronizer.scala#L89

And then here:
https://github.com/simplesteph/kafka-security-manager/blob/master/src/main/scala/com/github/simplesteph/ksm/AclSynchronizer.scala#L91-L104

but `errs` should not be empty, and therefore you should get:
https://github.com/simplesteph/kafka-security-manager/blob/master/src/main/scala/com/github/simplesteph/ksm/AclSynchronizer.scala#L100

But from the logs, it seems you don't get 
```
log.error("Exceptions while refreshing ACL source:")
```

which is really, weird, because errs should not be empty.... so I'm puzzled

michellewen1516 commented:
Thanks for this quick reply. 
But I don't think I got this log:
https://github.com/simplesteph/kafka-security-manager/blob/master/src/main/scala/com/github/simplesteph/ksm/source/BitbucketCloudSourceAcl.scala#L87

I think the below log is from: https://github.com/skinny-framework/skinny-framework/blob/master/http-client/src/main/scala/skinny/http/HTTP.scala

Feb 20 07:45:20 ip-172-31-11-158.ec2.internal kafka-security-manager[2341]: [2020-02-20 07:45:20,602] WARN Method(GET) https://api.bitbucket.org/2.0/repositories/ssimwave/kafka-acls/src/master/acl_test_1/acls.csv failed because connect timed out (skinny.http.HTTP$)
